lib/assertions.js

```diff
@@ -595,11 +595,16 @@ module.exports = function (expect) {
     expect.addAssertion('<function> to error [with] <any>', function (expect, subject, arg) {
         return expect(subject, 'to error').then(function (error) {
             expect.errorMode = 'nested';
-            if (error.isUnexpected && (typeof arg === 'string' || isRegExp(arg))) {
-                return expect(error, 'to have message', arg);
-            } else {
-                return expect(error, 'to satisfy', arg);
-            }
+            return expect.withError(function () {
+                if (error.isUnexpected && (typeof arg === 'string' || isRegExp(arg))) {
+                    return expect(error, 'to have message', arg);
+                } else {
+                    return expect(error, 'to satisfy', arg);
+                }
+            }, function (e) {
+                e._stack = error && error.stack;
+                throw e;
```

Comment:
Replacing

```js
                e._stack = error && error.stack;
                throw e;
```

with just

```js
                expect.fail(error);
```

seems to give exactly the same result. Would you prefer that?